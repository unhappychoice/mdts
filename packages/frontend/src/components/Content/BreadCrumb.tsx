import { Breadcrumbs, Link, Typography } from '@mui/material';
import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../../store/store';

interface BreadCrumbProps {
  onDirectorySelect?: (directoryPath: string) => void;
}

const BreadCrumb: React.FC<BreadCrumbProps> = ({ onDirectorySelect }) => {
  const { currentPath } = useSelector((state: RootState) => state.history);

  const pathSegments = currentPath ? currentPath.split('/').filter(segment => segment !== '') : [];

  return (
    <Breadcrumbs aria-label="breadcrumb" sx={{ mb: 4 }}>
      {pathSegments.map((segment, index) => {
        const isLast = index === pathSegments.length - 1;
        const path = pathSegments.slice(0, index + 1).join('/');
        return isLast ? (
          <Typography key={path} color="text.primary">
            {segment}
          </Typography>
        ) : (
          <Link key={path} color="inherit" href="#" onClick={() => onDirectorySelect && onDirectorySelect(path)}>
            {segment}
          </Link>
        );
      })}
    </Breadcrumbs>
  );
};

export default BreadCrumb;
