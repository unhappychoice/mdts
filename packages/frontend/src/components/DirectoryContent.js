"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_material_1 = require("@mui/icons-material");
const material_1 = require("@mui/material");
const FileTreeContext_1 = require("../contexts/FileTreeContext");
const filterFileTree = (tree, targetPath) => {
    if (targetPath === '') {
        const result = [];
        tree.forEach(item => {
            if (typeof item === 'string') {
                result.push(item.split('/').pop() || item);
            }
            else {
                const key = Object.keys(item)[0];
                const newObject = {};
                newObject[key.split('/').pop() || key] = item[key];
                result.push(newObject);
            }
        });
        return result;
    }
    const findChildren = (currentTree, pathSegments, currentSegmentIndex) => {
        if (currentSegmentIndex === pathSegments.length) {
            return currentTree;
        }
        const segment = pathSegments[currentSegmentIndex];
        for (const item of currentTree) {
            if (typeof item !== 'string') {
                const key = Object.keys(item)[0];
                const itemSegments = key.split('/');
                if (itemSegments[itemSegments.length - 1] === segment) {
                    const children = item[key];
                    if (Array.isArray(children)) {
                        return findChildren(children, pathSegments, currentSegmentIndex + 1);
                    }
                }
            }
        }
        return null;
    };
    const pathSegments = targetPath.split('/').filter(s => s !== '');
    const children = findChildren(tree, pathSegments, 0);
    if (children) {
        const result = [];
        children.forEach(item => {
            if (typeof item === 'string') {
                result.push(item.split('/').pop() || item);
            }
            else {
                const key = Object.keys(item)[0];
                const newObject = {};
                newObject[key.split('/').pop() || key] = item[key];
                result.push(newObject);
            }
        });
        return result;
    }
    return [];
};
const DirectoryContent = ({ selectedDirectoryPath, onFileSelect, onDirectorySelect, contentMode = 'fixed' }) => {
    const { fileTree: fullFileTree, loading, error } = (0, FileTreeContext_1.useFileTreeContext)();
    const fileTree = filterFileTree(fullFileTree, selectedDirectoryPath);
    const pathSegments = selectedDirectoryPath
        ? selectedDirectoryPath.split('/').filter(segment => segment !== '')
        : [];
    if (error)
        return (0, jsx_runtime_1.jsxs)("p", { children: ["Error: ", error] });
    const handleItemClick = (itemPath, isDirectory) => {
        if (isDirectory) {
            onDirectorySelect(itemPath);
        }
        else {
            onFileSelect(itemPath);
        }
    };
    return ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: Object.assign({ flexGrow: 1, p: 4, bgcolor: 'background.paper' }, (contentMode === 'fixed' && {
            maxWidth: '800px',
            margin: '0 auto',
        })), children: [selectedDirectoryPath && ((0, jsx_runtime_1.jsx)(material_1.Breadcrumbs, { "aria-label": "breadcrumb", sx: { mb: 4 }, children: pathSegments.map((segment, index) => {
                    const isLast = index === pathSegments.length - 1;
                    const path = pathSegments.slice(0, index + 1).join('/');
                    return isLast ? ((0, jsx_runtime_1.jsx)(material_1.Typography, { color: "text.primary", children: segment }, path)) : ((0, jsx_runtime_1.jsx)(material_1.Link, { color: "inherit", href: "#", onClick: () => onDirectorySelect && onDirectorySelect(path), children: segment }, path));
                }) })), (0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center', mb: 8 }, children: [(0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { sx: { mr: 2 }, color: "primary", fontSize: "large" }), (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "h4", gutterBottom: true, sx: { mb: 0 }, children: selectedDirectoryPath })] }), (0, jsx_runtime_1.jsx)(material_1.Divider, { sx: { paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' } }), loading && ((0, jsx_runtime_1.jsx)(material_1.Box, { sx: { display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }, children: (0, jsx_runtime_1.jsx)(material_1.CircularProgress, {}) })), (0, jsx_runtime_1.jsx)(material_1.Box, { mt: 4, children: fileTree && ((0, jsx_runtime_1.jsx)(material_1.List, { sx: { mr: -2, ml: -2 }, children: fileTree.map((item) => {
                        let name;
                        let itemPath;
                        let isDirectory;
                        if (typeof item === 'string') {
                            // It's a file
                            name = item;
                            itemPath = selectedDirectoryPath === '' ? item : `${selectedDirectoryPath}/${item}`;
                            isDirectory = false;
                        }
                        else {
                            // It's a directory object { [key: string]: FileTreeItem[] | string; }
                            const key = Object.keys(item)[0];
                            name = key;
                            itemPath = selectedDirectoryPath === '' ? key : `${selectedDirectoryPath}/${key}`;
                            isDirectory = true;
                        }
                        return ((0, jsx_runtime_1.jsxs)(material_1.ListItem, { button: true, onClick: () => handleItemClick(itemPath, isDirectory), children: [(0, jsx_runtime_1.jsx)(material_1.ListItemIcon, { sx: { minWidth: "38px" }, children: isDirectory ? (0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { color: "primary" }) : (0, jsx_runtime_1.jsx)(icons_material_1.ArticleOutlined, {}) }), (0, jsx_runtime_1.jsx)(material_1.ListItemText, { primary: name })] }, itemPath));
                    }) })) })] }));
};
exports.default = DirectoryContent;
