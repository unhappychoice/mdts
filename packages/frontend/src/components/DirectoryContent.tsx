import { ArticleOutlined, FolderOutlined } from '@mui/icons-material';
import {
  Box,
  Breadcrumbs,
  CircularProgress,
  Divider,
  Link,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Typography
} from '@mui/material';
import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../store/store';
import { selectFilteredFileTree } from '../store/slices/fileTreeSlice';

interface DirectoryContentProps {
  selectedDirectoryPath: string;
  onFileSelect: (filePath: string) => void;
  onDirectorySelect: (directoryPath: string) => void;
  contentMode?: 'fixed' | 'full';
}

const DirectoryContent: React.FC<DirectoryContentProps> = ({ selectedDirectoryPath, onFileSelect, onDirectorySelect, contentMode = 'fixed' }) => {
  const { fileTree: fullFileTree, loading, error } = useSelector((state: RootState) => state.fileTree);
  const fileTree = selectFilteredFileTree(fullFileTree, selectedDirectoryPath);

  const pathSegments = selectedDirectoryPath
    ? selectedDirectoryPath.split('/').filter(segment => segment !== '')
    : [];

  if (error) return <p>Error: {error}</p>;

  const handleItemClick = (itemPath: string, isDirectory: boolean) => {
    if (isDirectory) {
      onDirectorySelect(itemPath);
    } else {
      onFileSelect(itemPath);
    }
  };

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'fixed' && {
          maxWidth: '800px',
          margin: '0 auto',
        }),
      }}
    >
      {selectedDirectoryPath && (
        <Breadcrumbs aria-label="breadcrumb" sx={{ mb: 4 }}>
          {pathSegments.map((segment, index) => {
            const isLast = index === pathSegments.length - 1;
            const path = pathSegments.slice(0, index + 1).join('/');
            return isLast ? (
              <Typography key={path} color="text.primary">
                {segment}
              </Typography>
            ) : (
              <Link key={path} color="inherit" href="#" onClick={() => onDirectorySelect && onDirectorySelect(path)}>
                {segment}
              </Link>
            );
          })}
        </Breadcrumbs>
      )}
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 8 }}>
        <FolderOutlined sx={{ mr: 2 }} color="primary" fontSize="large"/>
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {selectedDirectoryPath}
        </Typography>
      </Box>
      <Divider sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }} />
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      )}
      <Box mt={4}>
        {fileTree && (
          <List sx={{ mr: -2, ml: -2 }}>
            {fileTree.map((item) => {
              let name: string;
              let itemPath: string;
              let isDirectory: boolean;

              if (typeof item === 'string') {
                // It's a file
                name = item;
                itemPath = selectedDirectoryPath === '' ? item : `${selectedDirectoryPath}/${item}`;
                isDirectory = false;
              } else {
                // It's a directory object { [key: string]: FileTreeItem[] | string; }
                const key = Object.keys(item)[0];
                name = key;
                itemPath = selectedDirectoryPath === '' ? key : `${selectedDirectoryPath}/${key}`;
                isDirectory = true;
              }

              return (
                <ListItem button key={itemPath} onClick={() => handleItemClick(itemPath, isDirectory)}>
                  <ListItemIcon sx={{ minWidth: "38px" }}>
                    {isDirectory ? <FolderOutlined color="primary" /> : <ArticleOutlined />}
                  </ListItemIcon>
                  <ListItemText primary={name} />
                </ListItem>
              );
            })}
          </List>
        )}
      </Box>
    </Box>
  );
};

export default DirectoryContent;
