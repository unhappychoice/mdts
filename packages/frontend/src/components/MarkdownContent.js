"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_material_1 = require("@mui/icons-material");
const material_1 = require("@mui/material");
const react_1 = require("react");
const useContent_1 = require("../hooks/apis/useContent");
const FileTreeContext_1 = require("../contexts/FileTreeContext");
const MarkdownPreview_1 = __importDefault(require("./MarkdownPreview"));
const Content = ({ selectedFilePath, contentMode = 'fixed', scrollToId, onDirectorySelect }) => {
    const [viewMode, setViewMode] = (0, react_1.useState)('preview');
    const { content, loading: contentLoading, error } = (0, useContent_1.useContent)(selectedFilePath);
    const { loading: fileTreeLoading } = (0, FileTreeContext_1.useFileTreeContext)();
    const loading = contentLoading || fileTreeLoading;
    (0, react_1.useEffect)(() => {
        if (scrollToId) {
            const element = document.getElementById(scrollToId);
            if (element) {
                element.scrollIntoView({ behavior: 'smooth', block: 'start' });
            }
        }
    }, [scrollToId]);
    const displayFileName = selectedFilePath
        ? selectedFilePath.split('/').pop()
        : loading ? '' : 'ðŸŽ‰ Welcome to mdts!';
    const pathSegments = selectedFilePath
        ? selectedFilePath.split('/').filter(segment => segment !== '')
        : [];
    if (error)
        return (0, jsx_runtime_1.jsxs)("p", { children: ["Error: ", error] });
    return ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: Object.assign({ flexGrow: 1, p: 4, bgcolor: 'background.paper' }, (contentMode === 'fixed' && {
            maxWidth: '800px',
            margin: '0 auto',
        })), children: [selectedFilePath && ((0, jsx_runtime_1.jsx)(material_1.Breadcrumbs, { "aria-label": "breadcrumb", sx: { mb: 4 }, children: pathSegments.map((segment, index) => {
                    const isLast = index === pathSegments.length - 1;
                    const path = pathSegments.slice(0, index + 1).join('/');
                    return isLast ? ((0, jsx_runtime_1.jsx)(material_1.Typography, { color: "text.primary", children: segment }, path)) : ((0, jsx_runtime_1.jsx)(material_1.Link, { color: "inherit", href: "#", onClick: () => onDirectorySelect && onDirectorySelect(path), children: segment }, path));
                }) })), (0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center', mb: 4 }, children: [(0, jsx_runtime_1.jsx)(icons_material_1.ArticleOutlined, { sx: { mr: 2 }, fontSize: "large" }), (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "h4", gutterBottom: true, sx: { mb: 0 }, children: displayFileName })] }), (0, jsx_runtime_1.jsx)(material_1.Box, { sx: { paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }, children: (0, jsx_runtime_1.jsxs)(material_1.Tabs, { value: viewMode, onChange: (event, newValue) => setViewMode(newValue), "aria-label": "view mode tabs", children: [(0, jsx_runtime_1.jsx)(material_1.Tab, { value: "preview", label: "Preview" }), (0, jsx_runtime_1.jsx)(material_1.Tab, { value: "raw", label: "Raw" })] }) }), loading ? ((0, jsx_runtime_1.jsx)(material_1.Box, { sx: { display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }, children: (0, jsx_runtime_1.jsx)(material_1.CircularProgress, {}) })) : viewMode === 'preview' ? ((0, jsx_runtime_1.jsx)(MarkdownPreview_1.default, { content: content })) : ((0, jsx_runtime_1.jsx)(material_1.Box, { component: "pre", sx: { whiteSpace: 'pre-wrap', p: 2, bgcolor: 'background.default', borderRadius: 1 }, children: content }))] }));
};
exports.default = Content;
