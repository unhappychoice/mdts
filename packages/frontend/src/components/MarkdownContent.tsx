import { ArticleOutlined } from '@mui/icons-material';
import { Box, Breadcrumbs, CircularProgress, Link, Tab, Tabs, Typography } from '@mui/material';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from '../store/store';
import { fetchContent } from '../store/slices/contentSlice';
import MarkdownPreview from './MarkdownPreview';

interface ContentProps {
  selectedFilePath: string | null;
  contentMode?: 'fixed' | 'full';
  scrollToId: string | null;
  onDirectorySelect?: (directoryPath: string) => void;
}

const Content: React.FC<ContentProps> = ({ selectedFilePath, contentMode = 'fixed', scrollToId, onDirectorySelect }) => {
  const [viewMode, setViewMode] = useState<'preview' | 'raw'>('preview');
  const dispatch = useDispatch<AppDispatch>();
  const { content, loading: contentLoading, error } = useSelector((state: RootState) => state.content);
  const { loading: fileTreeLoading } = useSelector((state: RootState) => state.fileTree);

  const loading = contentLoading || fileTreeLoading;

  useEffect(() => {
    dispatch(fetchContent(selectedFilePath));
  }, [dispatch, selectedFilePath]);

  useEffect(() => {
    if (scrollToId) {
      const element = document.getElementById(scrollToId);
      if (element) {
        element.scrollIntoView({ behavior: 'smooth', block: 'start' });
      }
    }
  }, [scrollToId]);

  const displayFileName = selectedFilePath
    ? selectedFilePath.split('/').pop()
    : loading ? '' : 'ðŸŽ‰ Welcome to mdts!';

  const pathSegments = selectedFilePath
    ? selectedFilePath.split('/').filter(segment => segment !== '')
    : [];

  if (error) return <p>Error: {error}</p>;

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'fixed' && {
          maxWidth: '800px',
          margin: '0 auto',
        }),
      }}
    >
      {selectedFilePath && (
        <Breadcrumbs aria-label="breadcrumb" sx={{ mb: 4 }}>
          {pathSegments.map((segment, index) => {
            const isLast = index === pathSegments.length - 1;
            const path = pathSegments.slice(0, index + 1).join('/');
            return isLast ? (
              <Typography key={path} color="text.primary">
                {segment}
              </Typography>
            ) : (
              <Link key={path} color="inherit" href="#" onClick={() => onDirectorySelect && onDirectorySelect(path)}>
                {segment}
              </Link>
            );
          })}
        </Breadcrumbs>
      )}
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 4 }}>
        <ArticleOutlined sx={{ mr: 2 }} fontSize="large" />
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {displayFileName}
        </Typography>
      </Box>
      <Box sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }}>
        <Tabs value={viewMode} onChange={(event, newValue) => setViewMode(newValue)} aria-label="view mode tabs">
          <Tab value="preview" label="Preview" />
          <Tab value="raw" label="Raw" />
        </Tabs>
      </Box>
      {loading ? (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      ) : viewMode === 'preview' ? (
        <MarkdownPreview content={content} selectedFilePath={selectedFilePath} />
      ) : (
        <Box component="pre" sx={{ whiteSpace: 'pre-wrap', p: 2, bgcolor: 'background.default', borderRadius: 1 }}>
          {content}
        </Box>
      )}
    </Box>
  );
};

export default Content;
