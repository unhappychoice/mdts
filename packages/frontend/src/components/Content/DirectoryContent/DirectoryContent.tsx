import { FolderOutlined } from '@mui/icons-material';
import { Box, CircularProgress, Divider, Typography } from '@mui/material';
import React, { useCallback } from 'react';
import { useSelector } from 'react-redux';
import { selectFilteredFileTree } from '../../../store/slices/fileTreeSlice';
import { RootState } from '../../../store/store';
import ErrorView from '../../ErrorView';
import BreadCrumb from '../BreadCrumb';
import { FileTreeList } from './FileTreeList';

interface DirectoryContentProps {
  onFileSelect: (filePath: string) => void;
  onDirectorySelect: (directoryPath: string) => void;
}

const DirectoryContent: React.FC<DirectoryContentProps> = ({ onFileSelect, onDirectorySelect }) => {
  const { currentPath } = useSelector((state: RootState) => state.history);
  const { contentMode } = useSelector((state: RootState) => state.appSetting);
  const { fileTree: fullFileTree, loading, error } = useSelector((state: RootState) => state.fileTree);

  const fileTree = selectFilteredFileTree(fullFileTree, currentPath);

  const handleItemClick = useCallback((itemPath: string, isDirectory: boolean) => {
    if (isDirectory) {
      onDirectorySelect(itemPath);
    } else {
      onFileSelect(itemPath);
    }
  }, [onDirectorySelect, onFileSelect]);

  if (error) {
    return <ErrorView error={error} />;
  }

  return (
    <Box
      sx={{
        width: '100%',
        minHeight: 'calc(100vh - 64px)',
        m: 0,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'compact' && {
          width: '800px',
          margin: '0 auto',
        })
      }}
    >
      <BreadCrumb onDirectorySelect={onDirectorySelect} />
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 4 }}>
        <FolderOutlined sx={{ mr: 2 }} color="primary" fontSize="large" />
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {currentPath}
        </Typography>
      </Box>
      <Divider sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }} />
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      )}
      <Box mt={4}>
        <FileTreeList fileTree={fileTree} handleItemClick={handleItemClick} />
      </Box>
    </Box>
  );
};

export default DirectoryContent;
