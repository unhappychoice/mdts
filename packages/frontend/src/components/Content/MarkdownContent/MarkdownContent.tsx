import { ArticleOutlined } from '@mui/icons-material';
import { Box, Chip, Typography, } from '@mui/material';
import React, { useEffect } from 'react';
import { useFrontmatter } from '../../../hooks/useFrontmatter';
import { useDispatch, useSelector } from 'react-redux';
import { fetchContent } from '../../../store/slices/contentSlice';
import { AppDispatch, RootState } from '../../../store/store';
import BreadCrumb from '../BreadCrumb';
import ErrorView from '../ErrorView';
import MarkdownContentTabs from './MarkdownContentTabs';
import MarkdownContentView from './MarkdownContentView';
import { useViewMode } from '../../../hooks/useViewMode';

interface MarkdownContentProps {
  scrollToId: string | null;
  onDirectorySelect?: (directoryPath: string) => void;
}

const MarkdownContent: React.FC<MarkdownContentProps> = ({ scrollToId, onDirectorySelect }) => {
  const dispatch = useDispatch<AppDispatch>();

  const { currentPath } = useSelector((state: RootState) => state.history);
  const { contentMode } = useSelector((state: RootState) => state.appSetting);
  const { content, loading: contentLoading, error } = useSelector((state: RootState) => state.content);
  const { loading: fileTreeLoading } = useSelector((state: RootState) => state.fileTree);

  const { frontmatter, markdownContent } = useFrontmatter(content);
  const viewMode = useViewMode();
  const loading = contentLoading || fileTreeLoading;

  useEffect(() => {
    dispatch(fetchContent(currentPath));
  }, [dispatch, currentPath]);

  useEffect(() => {
    if (scrollToId) {
      const element = document.getElementById(scrollToId);
      if (element) {
        element.scrollIntoView({ behavior: 'smooth', block: 'start' });
      }
    }
  }, [scrollToId, content]);
  const displayFileName = frontmatter.title
    ? String(frontmatter.title)
    : currentPath
      ? currentPath.split('/').pop()
      : loading
        ? ''
        : 'ðŸŽ‰ Welcome to mdts!';

  const hasFrontmatter = Object.keys(frontmatter).length > 0;

  if (error) {
    return <ErrorView error={error} />;
  }

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'fixed' && {
          maxWidth: '800px',
          margin: '0 auto',
        }),
      }}
    >
      <BreadCrumb onDirectorySelect={onDirectorySelect} />
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 4 }}>
        <ArticleOutlined sx={{ mr: 2 }} fontSize="large" />
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {displayFileName}
        </Typography>
      </Box>
      {frontmatter.tags && Array.isArray(frontmatter.tags) && frontmatter.tags.length > 0 && (
        <Box sx={{ display: 'flex', gap: 1, mb: 2 }}>
          {frontmatter.tags.map((tag) => (
            <Chip key={tag} label={tag} size="small" />
          ))}
        </Box>
      )}
      {frontmatter.categories && Array.isArray(frontmatter.categories) && frontmatter.categories.length > 0 && (
        <Box sx={{ display: 'flex', gap: 1, mb: 2 }}>
          {frontmatter.categories.map((category) => (
            <Chip key={category} label={category} size="small" />
          ))}
        </Box>
      )}
      <MarkdownContentTabs viewMode={viewMode} hasFrontmatter={hasFrontmatter} />
      <MarkdownContentView
        loading={loading}
        viewMode={viewMode}
        content={content}
        frontmatter={frontmatter}
        markdownContent={markdownContent}
      />
    </Box>
  );
};

export default MarkdownContent;
