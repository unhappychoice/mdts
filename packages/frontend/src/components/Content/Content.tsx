import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../../store/store';
import DirectoryContent from './DirectoryContent/DirectoryContent';
import MarkdownContent from './MarkdownContent/MarkdownContent';
import { Box } from '@mui/material';

interface ContentProps {
  onFileSelect: (path: string) => void;
  onDirectorySelect: (path: string) => void;
  scrollToId: string | null;
}

const Content: React.FC<ContentProps> = ({ onFileSelect, onDirectorySelect, scrollToId }) => {
  const { currentPath, isDirectory } = useSelector((state: RootState) => state.history);
  const { contentMode } = useSelector((state: RootState) => state.appSetting);

  const contentWidth = contentMode === 'compact' ? '800px' : '100%';

  return (
    <Box
      sx={{
        flexGrow: 1,
        overflowY: 'auto',
        display: 'flex',
        justifyContent: 'center',
        width: contentWidth,
        maxWidth: contentWidth,
        mx: 'auto',
      }}
    >
      {currentPath && isDirectory ? (
        <DirectoryContent onFileSelect={onFileSelect} onDirectorySelect={onDirectorySelect} />
      ) : (
        <MarkdownContent onDirectorySelect={onDirectorySelect} scrollToId={scrollToId} />
      )}
    </Box>
  );
};

export default Content;
