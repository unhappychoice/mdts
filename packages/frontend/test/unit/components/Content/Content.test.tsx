import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Content from '../../../../src/components/Content/Content';
import DirectoryContent from '../../../../src/components/Content/DirectoryContent/DirectoryContent';
import MarkdownContent from '../../../../src/components/Content/MarkdownContent/MarkdownContent';

const mockStore = configureStore([]);

jest.mock('../../../../src/components/Content/DirectoryContent/DirectoryContent');
jest.mock('../../../../src/components/Content/MarkdownContent/MarkdownContent');

describe('Content', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      history: {
        currentPath: null,
        isDirectory: false,
      },
      appSetting: {
        darkMode: 'auto',
        contentMode: 'compact',
        fileTreeOpen: true,
        outlineOpen: true,
      },
      content: {
        content: '',
        loading: false,
        error: null,
        scrollPosition: 0,
      },
    });
    DirectoryContent.mockClear();
    MarkdownContent.mockClear();

    (DirectoryContent as jest.Mock).mockImplementation(() => <div data-testid="mock-directory-content" />);
    (MarkdownContent as jest.Mock).mockImplementation(() => <div data-testid="mock-markdown-content" />);
  });

  test('renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Content onFileSelect={jest.fn()} onDirectorySelect={jest.fn()} scrollToId={null} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders MarkdownContent when currentPath is null', () => {
    render(
      <Provider store={store}>
        <Content onFileSelect={jest.fn()} onDirectorySelect={jest.fn()} scrollToId={null} />
      </Provider>
    );
    expect(screen.getByTestId('mock-markdown-content')).toBeInTheDocument();
    expect(screen.queryByTestId('mock-directory-content')).not.toBeInTheDocument();
  });

  test('renders MarkdownContent when currentPath is a file', () => {
    store = mockStore({
      history: {
        currentPath: 'path/to/file.md',
        isDirectory: false,
      },
      appSetting: {
        darkMode: 'auto',
        contentMode: 'compact',
        fileTreeOpen: true,
        outlineOpen: true,
      },
      content: {
        content: '',
        loading: false,
        error: null,
        scrollPosition: 0,
      },
    });
    render(
      <Provider store={store}>
        <Content onFileSelect={jest.fn()} onDirectorySelect={jest.fn()} scrollToId={null} />
      </Provider>
    );
    expect(screen.getByTestId('mock-markdown-content')).toBeInTheDocument();
    expect(screen.queryByTestId('mock-directory-content')).not.toBeInTheDocument();
  });

  test('renders DirectoryContent when currentPath is a directory', () => {
    store = mockStore({
      history: {
        currentPath: 'path/to/directory',
        isDirectory: true,
      },
      appSetting: {
        darkMode: 'auto',
        contentMode: 'compact',
        fileTreeOpen: true,
        outlineOpen: true,
      },
      content: {
        content: '',
        loading: false,
        error: null,
        scrollPosition: 0,
      },
    });
    render(
      <Provider store={store}>
        <Content onFileSelect={jest.fn()} onDirectorySelect={jest.fn()} scrollToId={null} />
      </Provider>
    );
    expect(screen.getByTestId('mock-directory-content')).toBeInTheDocument();
    expect(screen.queryByTestId('mock-markdown-content')).not.toBeInTheDocument();
  });

  test('passes correct props to MarkdownContent', () => {
    const onDirectorySelectMock = jest.fn();
    render(
      <Provider store={store}>
        <Content onFileSelect={jest.fn()} onDirectorySelect={onDirectorySelectMock} scrollToId="test-id" />
      </Provider>
    );
    expect(MarkdownContent).toHaveBeenCalledWith(
      expect.objectContaining({
        onDirectorySelect: onDirectorySelectMock,
        scrollToId: 'test-id',
      }),
      undefined
    );
  });

  test('passes correct props to DirectoryContent', () => {
    const onFileSelectMock = jest.fn();
    const onDirectorySelectMock = jest.fn();
    store = mockStore({
      history: {
        currentPath: 'path/to/directory',
        isDirectory: true,
      },
      appSetting: {
        darkMode: 'auto',
        contentMode: 'compact',
        fileTreeOpen: true,
        outlineOpen: true,
      },
      content: {
        content: '',
        loading: false,
        error: null,
        scrollPosition: 0,
      },
    });
    render(
      <Provider store={store}>
        <Content onFileSelect={onFileSelectMock} onDirectorySelect={onDirectorySelectMock} scrollToId={null} />
      </Provider>
    );
    expect(DirectoryContent).toHaveBeenCalledWith(
      expect.objectContaining({
        onFileSelect: onFileSelectMock,
        onDirectorySelect: onDirectorySelectMock,
      }),
      undefined
    );
  });
});
