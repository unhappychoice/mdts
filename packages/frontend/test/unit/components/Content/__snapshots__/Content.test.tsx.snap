// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Content renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1fthtn0"
  >
    <div
      class="custom-scrollbar MuiBox-root css-ynu778"
    >
      <div
        data-testid="mock-markdown-content"
      />
    </div>
  </div>
</DocumentFragment>
`;
